### PR TITLE
EN-2341 Added Component Persistence

### PR DESCRIPTION
**Pull request by rzsouza**

### CODE REVIEW

**Comment by entando-service** [comment]
All is well

**Comment by entando-service** [comment]
All is well

File: src/main/java/org/entando/entando/aps/system/util/AutoRollback.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package org.entando.entando.aps.system.util;
+
+import java.sql.*;
+
+public class AutoRollback implements AutoCloseable {
+
+    private Connection conn;
```
[the remark this answers is not in the document]
Comment (rzsouza) [reply src/main/java/org/entando/entando/aps/system/util/AutoRollback.java]: Fixed

File: src/main/java/org/entando/entando/aps/system/util/AutoSetAutoCommit.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package org.entando.entando.aps.system.util;
+
+import java.sql.*;
+
+public class AutoSetAutoCommit implements AutoCloseable {
+
+    private Connection conn;
+    private boolean originalAutoCommit;
```
[the remark this answers is not in the document]
Comment (rzsouza) [reply src/main/java/org/entando/entando/aps/system/util/AutoSetAutoCommit.java]: Fixed

File: src/main/java/org/entando/entando/web/digitalexchange/component/DEComponentControllerExtension.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+        Optional<DEComponentPackageFile> maybePackageFile = service.getPackageFile(componentId);
+
+        if (!maybePackageFile.isPresent()) return new ResponseEntity<>(HttpStatus.NOT_FOUND);
+
+        DEComponentPackageFile dePackageFile = maybePackageFile.get();
+
+        return ResponseEntity.ok()
+                .headers(createDownloadHeaders(dePackageFile.fileName))
+                .body(createDownloadBody(dePackageFile.inputStream));
```
[the remark this answers is not in the document]
Comment (rzsouza) [reply src/main/java/org/entando/entando/web/digitalexchange/component/DEComponentControllerExtension.java]: I replaced it by a InputStreamResource that will be automatically closed by Spring's  ResourceHttpMessageConverter

File: src/main/java/org/entando/entando/aps/system/init/portdb/DEComponentEntity.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    public static final String COL_ID = "id";
+    public static final String COL_NAME = "name";
+    public static final String COL_LAST_UPDATE = "lastupdate";
+    public static final String COL_COMPONENT_TYPE = "componenttype";
+    public static final String COL_DESCRIPTION = "description";
+
+
+    @DatabaseField(columnName = COL_ID,
+            dataType = DataType.STRING,
+            canBeNull = false, id = true)
```
[the remark this answers is not in the document]
Comment (rzsouza) [reply src/main/java/org/entando/entando/aps/system/init/portdb/DEComponentEntity.java]: Fixed and added validation for all String fields

**Pull request merged**